File: src/features/orderflow-edge-lab/types/conditions.ts
import type { Direction, InstrumentSymbol, SignalType, Timeframe } from "./common";

/** Mirrors engine/src/edge_lab/models/conditions.py::DetectedCondition */
export interface DetectedCondition {
  id: string;
  symbol: InstrumentSymbol;
  timeframe: Timeframe;
  barIndex: number;
  /** ISO 8601, UTC */
  timestamp: string;
  /** Key into the signal registry, e.g. 'poc_sweep', 'absorption' */
  signalType: SignalType;
  direction: Direction;
  /** Detector's own confidence (0-1), independent of Edge Score weight */
  strength: number;
  evidence: Record<string, number | string>;
}

File: engine/src/edge_lab/models/conditions.py
from __future__ import annotations

from datetime import datetime

from pydantic import Field

from edge_lab.config import InstrumentSymbol
from edge_lab.models.base import CamelModel
from edge_lab.models.enums import Direction, SignalType, Timeframe


class DetectedCondition(CamelModel):
    """One instance of a signal detector firing on a specific bar.

    This is the "what we detected" layer — a raw, pre-hypothesis observation.
    Whether a condition matters is decided later by a SetupDefinition's
    weighted rules, not by the detector itself.
    """

    id: str
    symbol: InstrumentSymbol
    timeframe: Timeframe
    bar_index: int = Field(ge=0)
    timestamp: datetime
    signal_type: SignalType = Field(
        description="Key into the signal registry, e.g. 'poc_sweep', 'absorption'"
    )
    direction: Direction
    strength: float = Field(
        ge=0, le=1, description="Detector's own confidence, independent of Edge Score weight"
    )
    evidence: dict[str, float | int | str] = Field(
        default_factory=dict,
        description="Structured explanation, e.g. {'swept_level': 4512.25, 'volume_multiple': 3.1}",
    )
